src/goto-symex/symex_dereference.cpp

```diff
@@ -389,4 +390,23 @@ void goto_symext::dereference(
   // dereferencing may introduce new symbol_exprt
   // (like __CPROVER_memory)
   state.rename(expr, ns, goto_symex_statet::L1);
+
+  // dereferencing is likely to introduce new member-of-if constructs --
+  // for example, "x->field" may have become "(x == &o1 ? o1 : o2).field"
+  // Simplify (which converts that to (x == &o1 ? o1.field : o2.field)) before
+  // applying field sensitivity, which can turn such field-of-symbol expressions
+  // into atomic SSA expressions, but would have to rewrite all of 'o1'
+  // otherwise.
+  // make the structure of the lhs as simple as possible to avoid,
+  // Similarly, (b ? s1 : s2).member=X results in
+  // (b ? s1 : s2)=(b ? s1 : s2) with member:=X and then
+  // s1=b ? ((b ? s1 : s2) with member:=X) : s1
+  // when all we need is
+  // s1=s1 with member:=X [and guard b]
+  // s2=s2 with member:=X [and guard !b]
+  do_simplify(expr);
```

Comment:
I'm curious to see what the impact on performance this change has.

Reply:
I will benchmark (and I hope @romainbrenguier can do his benchmarking as well) once the dependencies #2068 and #3770 are merged so that we have a precise analysis.

Reply:
This could maybe be made faster with a simplify method which would focus on the conversions that are mentioned above.

Reply:
That is certainly one option - are there particular patterns that affect Java? (@smowton maybe?)

Reply:
Update: this appears to be causing performance degradation; we may need to look into a specialised simplifier to achieve this.